### PR TITLE
Updating ContainerStorage Swagger

### PR DESCRIPTION
**Pull request by aarongalang**
Updating containerstorage to match Azure-ContainerStorageRP swagger

containerstorage folder should be identical to Swagger/Microsoft.ContainerStorage https://msazure.visualstudio.com/One/_git/Azure-ContainerStorageRP?path=/src/Swagger/Microsoft.ContainerStorage&version=GBmain

Includes:
- Field Validations

### CODE REVIEW

File: specification/containerstorage/ContainerStorage.Management/package-lock.json
Comment (mikeharder) [inline specification/containerstorage/ContainerStorage.Management/package-lock.json:1]: @aarongalang: Please delete this file.  You want to use the `package.json` and `package-lock.json` in the repo root.

**Review by mikeharder — CHANGES_REQUESTED** [review]

File: specification/containerstorage/ContainerStorage.Management/package-lock.json
Comment (aarongalang) [reply specification/containerstorage/ContainerStorage.Management/package-lock.json:1]: package-lock.json has been deleted.

File: specification/containerstorage/ContainerStorage.Management/.npmrc
Comment (mikeharder) [inline specification/containerstorage/ContainerStorage.Management/.npmrc:1]: @aarongalang: You will also need to delete this file

**Review by mikeharder — DISMISSED** [review]
Once `.npmrc` is deleted, you should be able to use the check named "specs - typespec - pr" to know when your spec is fully compliant.  You can also run the validation locally using the following command:

```
npx tsv specification/containerstorage/ContainerStorage.Management
```

**Comment by mikeharder** [comment]
@aarongalang: Pipeline "specs - typespec - pr" is working for your PR now.  It looks like you need to update your TypeSpec to be formatted correctly, and also update your autorest using the latest TypeSpec package versions (from the root `package-lock.json`).

You can stage these changes on your dev machine with a single command.  `tsv` is the command-line tool for TypeSpecValidation.

```
$ npx tsv specification/containerstorage/ContainerStorage.Management
```

**Comment by mikeharder** [comment]
> > @aarongalang
> >             
> >             Aaron Galang
> >             FTE: Pipeline "specs - typespec - pr" is working for your PR now. It looks like you need to update your TypeSpec to be formatted correctly, and also update your autorest using the latest TypeSpec package versions (from the root `package-lock.json`).
> > You can stage these changes on your dev machine with a single command. `tsv` is the command-line tool for TypeSpecValidation.
> > ```
> > $ npx tsv specification/containerstorage/ContainerStorage.Management
> > ```
> 
> I seem to be missing a dist folder to run the command. Is there a step I missed to get the TypeSpecValidation.js file? I tried rerouting it to use the one in TypeSpecValidation/src, but it didn't seem to work. image

You need to run `npm ci` in the repo root to build the tool and install dependencies.  Have you tried this yet?

**Comment by aarongalang** [comment]
@mikeharder how does npx tsv work? I've tried using it to format my changes, but while it did cause specs - typespec - pr check to pass, it erased the x-ms-example references, and caused the model validations check to fail. When I readded the example references, the specs-typespec-pr check failed again

Does it just generate the typespec again and compare them? Because we have had to make some manual changes to our containerstorage.json typespec

**Comment by mikeharder** [comment]
> @mikeharder Mike Harder FTE how does npx tsv work? I've tried using it to format my changes, but while it did cause specs - typespec - pr check to pass, it erased the x-ms-example references, and caused the model validations check to fail. When I readded the example references, the specs-typespec-pr check failed again
> 
> Does it just generate the typespec again and compare them? Because we have had to make some manual changes to our containerstorage.json typespec

The `tsv` tool currently does 4 things:

1. Verifies your spec is using the root `package.json`
2. Runs `tsp compile . --warn-as-error` on your spec
   1. Any compiler errors (or warnings promoted to errors) cause validation to fail
   2. Regenerates autorest (using the `@azure-tools/typespec-autorest` package)
3. Runs `tsp format` on your spec, which formats your source code per the repo config
4. Runs `git diff`.  Any changes files cause validation to fail.

If you are saying that you need to manually change your generated autorest, this is currently not supported for TypeSpec specs in this repo.  We require the checked-in autorest to match what is generated by the version of `@azure-tools/typespec-autorest` in the root package.json.

Why do you think you need to manually change your generated autorest?

**Comment by aarongalang** [comment]
> > @mikeharder Mike Harder FTE how does npx tsv work? I've tried using it to format my changes, but while it did cause specs - typespec - pr check to pass, it erased the x-ms-example references, and caused the model validations check to fail. When I readded the example references, the specs-typespec-pr check failed again
> > Does it just generate the typespec again and compare them? Because we have had to make some manual changes to our containerstorage.json typespec
> 
> The `tsv` tool currently does 4 things:
> 
> 1. Verifies your spec is using the root `package.json`
> 2. Runs `tsp compile . --warn-as-error` on your spec
>    
>    1. Any compiler errors (or warnings promoted to errors) cause validation to fail
>    2. Regenerates autorest (using the `@azure-tools/typespec-autorest` package)
> 3. Runs `tsp format` on your spec, which formats your source code per the repo config
> 4. Runs `git diff`.  Any changes files cause validation to fail.
> 
> If you are saying that you need to manually change your generated autorest, this is currently not supported for TypeSpec specs in this repo. We require the checked-in autorest to match what is generated by the version of `@azure-tools/typespec-autorest` in the root package.json.
> 
> Why do you think you need to manually change your generated autorest?

mostly to fix some minor LintDiff errors.
However, when I use npx tsv to generate the autorest, the examples and example references are not generated, causing XMS_EXAMPLE_NOTFOUND_ERRORs. But when I generate the examples (and as a result add x-ms-example references to the autorest), the specs-typespec-pr check fails... Am I missing something that's causing this problem?

on a side note:
when generating the autorest, should it always be done with npx tsv, rather than tsp compile?
should the autorest always be generated in the azure-rest-api-specs repo? or is it fine to generate it in our own repository (with our own package.json)?

**Comment by aarongalang** [comment]
There is an avocado error in the pr due to the default version not having path /subscriptions/{}/resourceGroups/{}/providers/Microsoft.ContainerStorage/pools/{}/volumeSnapshots

However, we have depreciated "volumeSnapshots" in favor of "snapshots". could the avocado-approved label added to suppress this error?

**Comment by aarongalang** [comment]
I've left in version 2023-03-01 for the sake of fixing breaking changes, however my team would like to remove this version. This is a test version and is not currently being used by anyone, so we don't want to have to support it.

Would it be fine to remove it in here or in a subsequent pr?

File: specification/containerstorage/resource-manager/Microsoft.ContainerStorage/preview/2023-07-01-preview/containerstorage.json
Comment (visingla-ms) [inline specification/containerstorage/resource-manager/Microsoft.ContainerStorage/preview/2023-07-01-preview/containerstorage.json:434]: >snapshots [](http://example.com/codeflow?start=126&length=9)

Existing customers who are using previous preview version, aren't these breaking changes to them? How are you ensuring the breaking experience for them?

Comment (croomes) [reply specification/containerstorage/resource-manager/Microsoft.ContainerStorage/preview/2023-07-01-preview/containerstorage.json:434]: There aren't any existing customers as the RP is still in development and it hasn't been made available anywhere public. We're hoping we can remove the `2023-03-01` version, so that our first publicly accessible release will use `2023-07-01`.

Comment (aarongalang) [reply specification/containerstorage/resource-manager/Microsoft.ContainerStorage/preview/2023-07-01-preview/containerstorage.json:431]: As Simon says, 2023-03-01 is a test version and isn't being used by customers, so we want to remove it, so we don't have to support an unused version. Is there any chance we can do so in this PR?

**Comment by mikeharder** [comment]
@konrad-jamrozik: @aarongalang claims he does not have permissions to remove the `ARMChangesRequested` label, as instructed by this comment:

https://github.com/Azure/azure-rest-api-specs/pull/24749#issuecomment-1650513357

Is there something he should do to get the necessary permissions to remove the label?

**Comment by croomes** [comment]
I've removed it for @aarongalang

File: specification/containerstorage/resource-manager/Microsoft.ContainerStorage/preview/2023-07-01-preview/containerstorage.json
Comment (visingla-ms) [reply specification/containerstorage/resource-manager/Microsoft.ContainerStorage/preview/2023-07-01-preview/containerstorage.json:431]: Thank you for sharing the details. You also have breaking changes review required label on your PR. Please get a breaking changes approval from Azure Breaking Changes Reviewers <azbreakchangereview@microsoft.com> and ARM can review the changes afterwards. Once you have the approval, please remove the "ARMChangesRequested" label from the PR to make the PR visible to the ARM reviewers.

**Comment by mikeharder** [comment]
> I've removed it for @aarongalang

@croomes: Are PR authors supposed to be able to remove the label themselves?

**Comment by aarongalang** [comment]
How do I fix the Service API Readiness Test?

**Comment by markcowl** [comment]
@aarongalang  The Staging checks are not required for check-in, you are currently blocked for needing an approval

**Comment by aarongalang** [comment]
> @aarongalang The Staging checks are not required for check-in, you are currently blocked for needing an approval

Where do I go to ask for approval?

**Comment by konrad-jamrozik** [comment]
> > @aarongalang The Staging checks are not required for check-in, you are currently blocked for needing an approval
> 
> Where do I go to ask for approval?

Please refer to this diagram:

- https://github.com/Azure/azure-rest-api-specs/pull/25071

Specifically, you need to add the `MergeRequested` label

**Review by croomes — APPROVED** [review]

**Pull request merged**
